library/modules/Gui.cpp

```diff
@@ -1360,6 +1368,100 @@ DFHACK_EXPORT void Gui::writeToGamelog(std::string message)
     fseed.close();
 }
 
+namespace
+{   // Utility functions for reports
+    bool parseReportString(std::vector<std::string> *out, const std::string &str, size_t line_length = 73)
+    {   // parse a string into output strings like DF does for reports
+        if (str.empty() || line_length == 0)
+            return false;
+
+        string parsed;
+        size_t i = 0;
+
+        do
+        {
+            if (str[i] == '&') // escape character
+            {
+                i++; // ignore the '&' itself
+                if (i >= str.length())
+                    break;
+
+                if (str[i] == 'r') // "&r" adds a blank line
+                {
+                    word_wrap(out, parsed, line_length, false/*, true*/);
+                    out->push_back(" "); // DF adds a line with a space for some reason
+                    parsed.clear();
+                }
+                else if (str[i] == '&') // "&&" is '&'
+                    parsed.push_back('&');
+                // else next char is ignored
+            }
+            else
+                parsed.push_back(str[i]);
+        }
+        while (++i < str.length());
+
+        if (parsed.length())
+            word_wrap(out, parsed, line_length, false/*, true*/);
+
+        return true;
+    }
+
+    bool recent_report(df::unit *unit, df::unit_report_type slot)
+    {
+        if (unit && !unit->reports.log[slot].empty() &&
+            *df::global::cur_year == unit->reports.last_year[slot] &&
+            (*df::global::cur_year_tick - unit->reports.last_year_tick[slot]) <= 500)
+        {
+            return true;
+        }
+        return false;
+    }
+
+    bool recent_report_any(df::unit *unit)
+    {
+        FOR_ENUM_ITEMS(unit_report_type, slot)
+        {
+            if (recent_report(unit, slot))
+                return true;
+        }
+        return false;
+    }
+
+    void delete_old_reports()
+    {
+        auto &reports = world->status.reports;
+        while (reports.size() > 3000)
+        {
+            if (reports[0] != NULL)
+            {
+                if (reports[0]->flags.bits.announcement)
+                    erase_from_vector(world->status.announcements, &df::report::id, reports[0]->id);
+                delete reports[0];
+            }
+            reports.erase(reports.begin());
+        }
+    }
```

Comment:
how often do we expect to run this code? This is O(M*N) where M is the number of erased items and N is the number of items in the vector (which is > 3000). It seems like it could be brought down to O(N) fairly easily by batch removing elements.

Reply:
``delete_old_reports`` is equivalent to a loop DF runs at the end of each announcement (or report) made. (The part run when ``reports[0] != NULL`` is DF's report destructor.) Technically, we could just not do this, and DF would handle it the next time the game does one of its own announcements. We'd be adding more iterations for DF to do in a single tick, however.

The number of erased items (M) should be equal to the number of report entries (lines of text) we just added, assuming we're keeping on top of things. That's most likely going to be a number < 5, which isn't bad.

So we could do slightly better, but this is what DF is doing all the time.